### ISSUE TITLE
Add additional information for lab/test data

### ISSUE THREAD

**Issue by acagreen17** [opensafely-core/cohort-extractor#601]
This request has been brought about through an external OpenSAFELY project but if we are increasingly using test values it is likely that the additional information will be useful going forward.

**Use-case from our external OpenSAFELY user:**
_We are currently using OpenSafely to assist with COVID-19 recovery planning in the area of colorectal cancer. Quantitative faecal immunochemical testing (FIT) has been more widely deployed in primary care during the pandemic period. We aim to understand how its use has impacted colorectal cancer diagnostic pathways and diagnoses. At present, laboratory data in the primary care recorded can be queried using the with_these_clinical_events, returning dates and numeric data. However, for some tests like FIT, the result is often reported using an inequality sign (e.g. <10 or >200). This is important - <10 clearly has a different meaning to 10. For some results, there may just be a normal/abnormal result, which will be missed altogether by the conversion to a number._
 
_I would therefore propose that the following are made available via the study definition interface:_
_1. Any inequality sign if present in the original data (‘>’, ‘<’ or none)_
_2. An abnormal/normal flag (with some indication as to whether this flag is simply not present; i.e. a TRUE/FALSE/missing value rather than a simple Boolean value)_
_3. As an alternative to (2), and depending on how the data are stored, a normal/low/critically low/high/critically high/missing value_
_4. The reference range_

To back up this use-case, there are some histograms of the FIT results, which do demonstrate the disproportionately large number of 0s/10s (numbers in the 000's). It’s worth noting that the common values vary by STP. In some STPs it’s 4, others 7, others 10. Some have lots of zeros (which is the default for missing values, but could potentially also represent actually a textual result?) Note, these can be re-formatted and redacted as needed.

**Comment by evansd** [comment]
I think we have the raw data to support something like this in TPP, via the `CodedEventRange` table:

ColumnName | ColumnType | MaxLength | Precision | Scale | IsNullable
-- | -- | -- | -- | -- | --
Patient_ID | int | 4 | 10 | 0 | False
CodedEvent_ID | bigint | 8 | 19 | 0 | False
CodedEventRange_ID | bigint | 8 | 19 | 0 | False
Consultation_ID | bigint | 8 | 19 | 0 | False
LowerBound | real | 4 | 24 | 0 | False
UpperBound | real | 4 | 24 | 0 | False
Comparator | tinyint | 1 | 3 | 0 | False

However I don't think we have anything equivalent for the EMIS `Observations` table. It looks like we get just a single numeric value (`value_pq_1`).

It's possible @inglesp might know if there's additional data in EMIS which we're not yet representing in our version of the schema though.

**Comment by inglesp** [comment]
I'm not aware of anything like this in the EMIS data we have access to.

**Comment by acagreen17** [comment]
@sebbacon do you have a rough idea of then this might be done? (Not trying to press, but would like to update our external OpenSAFELY user).

**Comment by sebbacon** [comment]
No - but am having a triaging meeting with @beckylumbard this afternoon!

**Comment by BeckyLumbard** [comment]
Out of the items marked as priority I think this data edit is the next priority item that is ready to go and has an external collaborator on a co-pilot project....Seb suggested this item can be scheduled in now!

**Comment by sebbacon** [comment]
Notes for whoever picks this up:
* We should ask EMIS if they can make the data available, though it might take a while. I have kicked this off in Slack and will be having a call with their PM soon (hopefully this week) 
* I did a lot of work on pathology data a couple of years ago, come to me for any domain-specific queries.
  * Note that pathology labs usually send a message to the EHR with the value; the reference ranges used; and an abnormal/normal (or too-low/too-high) indicator. Ideally we'd get (and return) all of these. Note that these won't necessarily agree with each other! But also note that the indicator would usually be what the GP looks at, and therefore makes clinical decisions on, so is the "right" thing to use.
  * Also note that data may not be normalised between labs, so the same test may have different possible values for flags when returned by different labs. My guess is that EMIS normalise this already
* It sounds like neither EMIS nor TPP provide a normal/abnormal flag at present. We should check with EMIS if they could make this available.  It's possible the flag makes its way to a different coding table. If not, then someone would have to impute this from the supplied values and reference ranges.
  * In the case that abnormal/normal is not made available, we could consider imputing normal/abnormal for the user (rather than leaving it up to them - though we should obviously discuss the correct implementation with the current users up front!), because there are decisions they'd have to make when doing it themselves, which perhaps are better handled in our (tested, documented, and therefore consistent) code. For example, don't be surprised if there are values like `<10` for a measurand whose reference range has a lower bound of `5`. Or what to do with nulls.
    * Of course, it may be that one backend does supply the abnormal flag from the underlying data, and another doesn't so it has to be imputed; in that case, we'd want our function to be able to indicate (or specify) which implementation was used for the provided value

**Comment by sebbacon** [comment]
Just to note - in EMIS the "effective date" is computed from a number of the other date fields. For our purposes, we are likely to be interested in the "specimen taken date" most of the time, and the "report received date" (when the report was linked to the patient's care record). So we've requested those two in EMIS.

**Comment by sebbacon** [comment]
For reference, we _might_ expect this data to land in EMIS in 2 - 4 weeks. Worth bearing in mind and consulting with our stakeholders (like @acagreen17) on urgency.

**Comment by acagreen17** [comment]
If there was any way to get this done sooner that would be ideal as our external user has been waiting a while and their project is at a bit of a standstill now. Do we need the data in EMIS to be able to use the data from TPP?

**Comment by sebbacon** [comment]
No. It's just a fair bit more work to do them at different times.

**Comment by acagreen17** [comment]
Okay fab! (But sorry for the extra work). Do you know what sort of time frame we're looking at for this to be useable?

**Comment by sebbacon** [comment]
Not yet, this is top of the "todo" list but some other things are in progress. I expect someone will pick it up before mid-next-week and will be able to provide an estimate then.

For reference, EMIS have suggested that 18th October is a realistic target date for data flowing to them.

**Comment by evansd** [comment]
### Notes / Questions

The available `Comparator` values in the TPP `CodedEventRange` table are:
| Value | Meaning |
|-------|---------|
| 3     | ~       |
| 4     | =       |
| 5     | >=      |
| 6     | >       |
| 7     | <       |
| 8     | <=      |

I can't immediately see a pattern in how these relate to the `LowerBound` and `UpperBound` values. Sometimes they're both -1. Often they both have different positive values and I don't know how to interpret e.g. `10 = 30`.

How does the `CodedEvent_SNOMED` table fit in here? There's no equivalent range table.

**Comment by sebbacon** [comment]
Here's my guess. The _value_ is in `CodedEvent_SNOMED` - say it's `22`; then you cross-reference `CodedEventRange` and find that the `Comparator` is `>`, so now you know it's 23 or more; and the lower bound is 10 and the upper bound is 20, so you know it's out of range.

Sometimes there is no reference range, so probably `-1` means "no upper/lower bound"; so there is no concept of out of range in that test.

Sometimes in OpenPathology you'd find a value of say `>10` and a reference range of `1-15`, meaning you've no idea if it's in range or not.

I'd say the best way to get to the bottom of this is to set up a call with Chris and @HelenCEBM (as our OpenPathology specialist) and get your questions answered that way.

**Comment by evansd** [comment]
Thanks @sebbacon. Just come independently to the same conclusion in discussion with @CarolineMorton, which is encouraging. Will obviously confirm with Chris though.

The SNOMED thing is a separate issue though I think. `CodedEventRange` has an FK to `CodedEvent` but (AFAICT) no equivalent link to `CodedEvent_SNOMED` so it's possible we'll have to say "no ref ranges if you're using SNOMED codes". But again, will confirm with Chris.
